janus_server/src/datastore.rs

```diff
@@ -0,0 +1,318 @@
+        let input_shares: Vec<HpkeCiphertext> =
+            decode_u16_items(&(), &mut Cursor::new(&encoded_input_shares))?;
+
+        Ok(Report {
+            task_id,
+            nonce: Nonce {
+                time: nonce_time,
+                rand: nonce_rand as u64,
+            },
+            extensions,
+            encrypted_input_shares: input_shares,
+        })
+    }
+
+    /// put_client_report stores a client report.
+    pub async fn put_client_report(&self, report: &Report) -> Result<i64, Error> {
+        let nonce_time = report.nonce.time.as_naive_date_time();
+        let nonce_rand = report.nonce.rand as i64;
+
+        let mut encoded_extensions = Vec::new();
+        encode_u16_items(&mut encoded_extensions, &(), &report.extensions);
+
+        let mut encoded_input_shares = Vec::new();
+        encode_u16_items(
+            &mut encoded_input_shares,
+            &(),
+            &report.encrypted_input_shares,
+        );
+
+        let row = self.tx.query_one(
+            "INSERT INTO client_reports (task_id, nonce_time, nonce_rand, extensions, input_shares)
+            VALUES ($1, $2, $3, $4, $5) RETURNING (id)",
+            &[
+                /* task_id */      &&report.task_id.0[..],
+                /* nonce_time */   &nonce_time,
+                /* nonce_rand */   &nonce_rand,
+                /* extensions */   &encoded_extensions,
+                /* input_shares */ &encoded_input_shares,
+            ]
+        ).await?;
+        Ok(row.get("id"))
+    }
+}
+
+fn single_row(rows: Vec<Row>) -> Result<Row, Error> {
+    match rows.len() {
+        0 => Err(Error::NotFound),
+        1 => Ok(rows.into_iter().next().unwrap()),
+        _ => Err(Error::TooManyRows),
+    }
+}
+
+/// Error represents a datastore-level error.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// An error from the underlying database library.
+    #[error("DB error: {0}")]
+    Db(#[from] tokio_postgres::Error),
+    /// An entity requested from the datastore was not found.
+    #[error("not found in database")]
+    NotFound,
+    /// A query that was expected to return at most one row unexpectedly returned more than one row.
+    #[error("multiple rows returned where only one row expected")]
+    TooManyRows,
+    /// An error from decoding a value stored encoded in the underlying database.
+    #[error("decoding error: {0}")]
+    DecodeError(#[from] CodecError),
+    /// An arbitrary error returned from the user callback; unrelated to DB internals. This error
+    /// will never be generated by the datastore library itself.
+    #[error(transparent)]
+    User(#[from] Box<dyn std::error::Error + Send + Sync>),
+}
+
+impl Error {
+    // is_serialization_failure determines if a given error corresponds to a Postgres
+    // "serialization" failure, which requires the entire transaction to be aborted & retried from
+    // the beginning per https://www.postgresql.org/docs/9.5/transaction-iso.html.
```

Comment:
In the test below, we use PostgreSQL version 14, but this links to docs for 9.5 which is unsupported.